packages/typescript/src/index.ts

```diff
@@ -42,6 +46,9 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
           resolveModule,
           parsedOptions,
           writeFile(fileName, data) {
+            if (parsedOptions.options.composite || parsedOptions.options.incremental) {
+              tsCache.cacheCode(fileName, data);
```

Comment:
Is there a Typescript API available for this functionality? I find it odd that these flags require us to create our own cache implementation.

Reply:
Well, the issue is this: `writeFile` overwrites the default typescript output.
  
  So, normally: 

  [fs] main.ts => tsc => `writeFile` => [fs] main.js

this plugin changes the `writeFile` to work like this:

  [fs] main.ts => tsc => `writeFile` => [emittedFiles] main.js

   When dealing with an already compiled file that has not changed - according to the `tsbuildinfo` - `tsc` will not write the file, since it's already on disk. 

  With the plugin - during an incremental build - the first time we run rollup we will emit the files correctly into the next step of the pipeline - correctly transpiled into javascript. The second time we run rollup the plugin will not see the file in the `emittedFiles` - since it hasn't changed and `tsc` didn't call `writeFile` - and it will pass the un-transpiled typescript file to the next step of the pipeline. This will fail, since it isn't  a javascript file.

  This PR creates a local cache of the files emitted by tsc during incremental runs, the next time rollup runs the transpiled files will be picked up from the cache and sent forward.

  In other words: yes - typescript has a cache: the transpiled files, but we are not saving them - since we only want to transpile on the fly and move the on with our lives. This will break when using incremental builds.

  I hope I have clarified a bit more than I have confused.

Reply:
Hello everyone, sorry for jumping into the conversation, but I wanted to highlight a few questions/concernes regarding the cache implementation:
- there is no **process awareness** in default caching destination: can be **easy to mitigate** by **using separate cache paths**/process or build. This can be a major problem when dealing with multiple concurrent builds and would highly suggest **documenting the edge case**.
- TSCache works on **file names only** and is **unawareness of contents changes**: can be **easy to mitigate** by **deleting the cache folder** before new builds, but beats the purpose in my view. One of incremental's advantages is lowering the "cold" build (as opposed to watch mode) duration, and deletion of cache will negate that particular advantage. I would highly suggest using hashes based on file's contents or at least file stats (mtime, ctime etc) to mittigate the issue, this would also work well with common use cases such as: changing a branch, pull-ing from version control etc.

Please take into consideration that my last point could be muted, by the typescript compiler determining changes and emitting new files (this should be the expected behavior, and emitted files have precedence over cached versions in current implementation); but I wanted to raise awareness.